gems/smithy-schema/lib/smithy-schema/document/serializer.rb

```diff
@@ -0,0 +1,283 @@
+# frozen_string_literal: true
+
+require 'base64'
+require 'time'
+
+module Smithy
+  module Schema
+    module Document
+      # Serializes data into a document data.
+      class Serializer
+        include Shapes
+
+        # @param [TypeRegistry] type_registry Used to find shape based on
+        #  on document discriminator.
+        def initialize(type_registry)
+          @type_registry = type_registry
+        end
+
+        # Create document data from various input data formats
+        # @param [Object] data Input data can be: Ruby objects, instance of a runtime shape or a
+        #  JSON response with type discriminator.
+        # @return [Data] document data
+        #
+        # @example Ruby objects as input
+        #   # create serializer with an existing type registry
+        #   serializer = Smithy::Schema::Document::Serializer(type_registry)
+        #
+        #   # ruby objects as input
+        #   serializer.create_document("some document")
+        #   # => {"foo" => "bar"}
+        # @example Runtime shape as input
+        #   # create serializer with an existing type registry
+        #   serializer = Smithy::Schema::Document::Serializer(type_registry)
+        #
+        #   # given the following runtime shape
+        #   runtime_shape = some_structure.new.type(some_data)
+        #   # => #<struct SampleService::Types::Structure...>
+        #
+        #   serializer.create_document(runtime_shape)
+        #   # => an instance of Smithy::Schema::Document::Data
+        # @example JSON data
+        #   # create serializer with an existing type registry
+        #   serializer = Smithy::Schema::Document::Serializer(type_registry)
+        #
+        #   # given the following json data
+        #   parsed_json = {
+        #     "__type" => "smithy.ruby.tests#Structure",
+        #     "string" => "hello"
+        #   }
+        #
+        #   document = serializer.create_document(parsed_json)
+        #   # => an instance of Smithy::Schema::Document::Data
+        #   document.discriminator
+        #   # => "smithy.ruby.tests#Structure"
+        def create_document(data)
```

Comment:
This feels out of place - this seems like it would be Document.new (as a class method or as initialize).